File: src/deadline/client/api/_telemetry.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

import atexit
from functools import lru_cache, wraps
import json
import logging
import os
import platform
import uuid
import random
import sys
import time

from botocore.config import Config as BotocoreConfig
from configparser import ConfigParser
from dataclasses import asdict, dataclass, field
from datetime import datetime
from queue import Queue, Full
from threading import Thread
from typing import Any, Callable, Dict, Optional, TypeVar, cast
from urllib import request, error

from ...job_attachments.progress_tracker import SummaryStatistics

from ._session import (
    get_monitor_id,
    get_user_and_identity_store_id,
    get_boto3_client,
    get_boto3_session,
)
from ._stack_trace_sanitizer import sanitize_exception
from ..config import config_file
from .. import version

__cached_telemetry_clients: Dict[str, "TelemetryClient"] = {}

logger = logging.getLogger(__name__)


# Generic function return type.
F = TypeVar("F", bound=Callable[..., Any])


def _swallow_exceptions(func: F) -> F:
    """Decorator that catches all exceptions in telemetry functions to prevent
    telemetry issues from affecting the main application flow."""

    @wraps(func)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        try:
            return func(*args, **kwargs)
        except Exception:
            logger.debug(
                "Swallowed exception in telemetry function %s", func.__name__, exc_info=True
            )
            return None

    return cast(F, wrapper)


def get_deadline_endpoint_url(
    config: Optional[ConfigParser] = None,
) -> str:
    # Use boto3's built-in logic to get the correct endpoint URL
    client = get_boto3_client("deadline", config=config)
    return client.meta.endpoint_url


@dataclass
class TelemetryEvent:
    """Base class for telemetry events"""

    event_type: str = "com.amazon.rum.deadline.uncategorized"
    event_details: Dict[str, Any] = field(default_factory=dict)


class TelemetryClient:
    """
    Sends telemetry events periodically to the Deadline Cloud telemetry service.

    This client holds a queue of events which is written to synchronously, and processed
    asynchronously, where events are sent in the background, so that it does not slow
    down user interactivity.

    Telemetry events contain non-personally-identifiable information that helps us
    understand how users interact with our software so we know what features our
    customers use, and/or what existing pain points are.

    Data is aggregated across a session ID (a UUID created at runtime), used to mark every
    telemetry event for the lifetime of the application), and a 'telemetry identifier' (a
    UUID recorded in the configuration file), to aggregate data across multiple application
    lifetimes on the same machine.

    Telemetry collection can be opted-out of by running:
    'deadline config set "telemetry.opt_out" true' or setting the environment variable
    'DEADLINE_CLOUD_TELEMETRY_OPT_OUT=true'
    """

    # Used for backing off requests if we encounter errors from the service.
    # See https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
    MAX_QUEUE_SIZE = 25
    BASE_TIME = 0.5
    MAX_BACKOFF_SECONDS = 10  # The maximum amount of time to wait between retries
    MAX_RETRY_ATTEMPTS = 4

    ENDPOINT_PREFIX = "management."

    def __init__(
        self,
        package_name: str,
        package_ver: str,
        config: Optional[ConfigParser] = None,
    ):
        # Instance-level dicts so every TelemetryClient has its own state (avoid
        # the mutable-class-attribute pitfall where updates would be shared by
        # all instances).
        self._common_details: Dict[str, Any] = {}
        self._system_metadata: Dict[str, Any] = {}

        self._initialized: bool = False
        self.package_name = package_name
        self.package_ver = ".".join(package_ver.split(".")[:3])

        # IDs for this session
        self.session_id: str = str(uuid.uuid4())
        try:
            self.telemetry_id: str = self._get_telemetry_identifier(config=config)
        except Exception:
            logger.debug("Swallowed exception in telemetry __init__", exc_info=True)
            self.telemetry_id = str(uuid.uuid4())
        # If a different base package is provided, include info from this library as supplementary info
        if package_name != "deadline-cloud-library":
            self._common_details["deadline-cloud-version"] = version
        try:
            self._system_metadata = self._get_system_metadata(config=config)
        except Exception:
            logger.debug("Swallowed exception in telemetry __init__", exc_info=True)
            self._system_metadata = {}
        self.set_opt_out(config=config)
        self.initialize(config=config)

    @_swallow_exceptions
    def set_opt_out(self, config: Optional[ConfigParser] = None) -> None:
        """
        Checks whether telemetry has been opted out by checking the DEADLINE_CLOUD_TELEMETRY_OPT_OUT
        environment variable and the 'telemetry.opt_out' config file setting.
        Note the environment variable supersedes the config file setting.
        """
        env_var_value = os.environ.get("DEADLINE_CLOUD_TELEMETRY_OPT_OUT")
        if env_var_value:
            self.telemetry_opted_out = env_var_value in config_file._TRUE_VALUES
        else:
            self.telemetry_opted_out = config_file.str2bool(
                config_file.get_setting("telemetry.opt_out", config=config)
            )
        logger.info(
            "Deadline Cloud telemetry is "
            + ("not enabled." if self.telemetry_opted_out else "enabled.")
        )

    @_swallow_exceptions
    def initialize(self, config: Optional[ConfigParser] = None) -> None:
        """
        Starts up the telemetry background thread after getting settings from the boto3 client.
        Note that if this is called before boto3 is successfully configured / initialized,
        an error can be raised. In that case we silently fail and don't mark the client as
        initialized.
        """
        if self.telemetry_opted_out:
            return

        self.endpoint: str = self._get_prefixed_endpoint(
            f"{get_deadline_endpoint_url(config=config)}/2023-10-12/telemetry",
            TelemetryClient.ENDPOINT_PREFIX,
        )

        # Some environments might not have SSL, so we'll use the vendored botocore SSL context
        from botocore.httpsession import create_urllib3_context, get_cert_path

        self._urllib3_context = create_urllib3_context()
        self._urllib3_context.load_verify_locations(cafile=get_cert_path(True))

        user_id, _ = get_user_and_identity_store_id(config=config)
        if user_id:
            self._system_metadata["user_id"] = user_id

        monitor_id: Optional[str] = get_monitor_id(config=config)
        if monitor_id:
            self._system_metadata["monitor_id"] = monitor_id

        self._initialized = True
        self._start_threads()

    def record_error_with_trace(
        self,
        exc: BaseException,
        exception_scope: str,
        extra_details: Optional[dict] = None,
        from_gui: bool = False,
    ) -> None:
        event_details: dict = {
            "exception_type": type(exc).__qualname__,
            "exception_scope": exception_scope,
            "stack_trace": sanitize_exception(exc),
        }
        if extra_details:
            event_details.update(extra_details)

        self.record_event(
            event_type="com.amazon.rum.deadline.error",
            event_details=event_details,
            from_gui=from_gui,
        )

    @property
    def is_initialized(self) -> bool:
        return self._initialized

    def _get_prefixed_endpoint(self, endpoint: str, prefix: str) -> str:
        """Insert the prefix right after 'https://'"""
        if endpoint.startswith("https://"):
            prefixed_endpoint = endpoint[:8] + prefix + endpoint[8:]
            return prefixed_endpoint
        return endpoint

    def _get_telemetry_identifier(self, config: Optional[ConfigParser] = None):
        identifier = config_file.get_setting("telemetry.identifier", config=config)
        try:
            uuid.UUID(identifier, version=4)
        except ValueError:  # Thrown if the user_id isn't in UUID4 format
            identifier = str(uuid.uuid4())
            config_file.set_setting("telemetry.identifier", identifier)
        return identifier

    def _start_threads(self) -> None:
        """Set up background threads for shutdown checking and request sending"""
        self.event_queue: Queue[Optional[TelemetryEvent]] = Queue(
            maxsize=TelemetryClient.MAX_QUEUE_SIZE
        )
        atexit.register(self._exit_cleanly)
        self.processing_thread: Thread = Thread(
            target=self._process_event_queue_thread, daemon=True
        )
        self.processing_thread.start()

    def _get_system_metadata(self, config: Optional[ConfigParser]) -> Dict[str, Any]:
        """
        Builds up a dict of non-identifiable metadata about the system environment.

        This will be used in the Rum event metadata, which has a limit of 10 unique values.
        """
        platform_info = platform.uname()
        metadata: Dict[str, Any] = {
            "service": self.package_name,
            "version": self.package_ver,
            "python_version": platform.python_version(),
            "osName": "macOS" if platform_info.system == "Darwin" else platform_info.system,
            "osVersion": platform_info.release,
        }

        return metadata

    @_swallow_exceptions
    def _exit_cleanly(self):
        try:
            self.event_queue.put_nowait(None)
        except Full:
            # If the queue is full, it may mean the telemetry processing thread has already joined
            # since it is daemon and the Python runtime will shut it down on exit.
            # Ignore the error, since this is a best-effort cleanup.
            pass
        self.processing_thread.join()

    def _send_request(self, req: request.Request) -> None:
        attempts = 0
        success = False
        while not success:
            try:
                with request.urlopen(req, context=self._urllib3_context):
                    logger.debug("Successfully sent telemetry.")
                    success = True
            except error.HTTPError as httpe:
                if httpe.code == 429 or httpe.code == 500:
                    logger.debug(f"Error received from service. Waiting to retry: {str(httpe)}")

                    attempts += 1
                    if attempts >= TelemetryClient.MAX_RETRY_ATTEMPTS:
                        raise Exception("Max retries reached sending telemetry")

                    backoff_sleep = random.uniform(
                        0,
                        min(
                            TelemetryClient.MAX_BACKOFF_SECONDS,
                            TelemetryClient.BASE_TIME * 2**attempts,
                        ),
                    )
                    time.sleep(backoff_sleep)
                else:  # Reraise any exceptions we didn't expect
                    raise

    def _process_event_queue_thread(self):
        """Background thread for processing the telemetry event data queue and sending telemetry requests."""
        # Resolve the AWS account ID once on this background thread so callers
        # of record_event() are never blocked (e.g. by a slow STS timeout on a
        # restricted network). The resolved value is stored on _common_details
        # and merged into every event's payload below, so events enqueued
        # before resolution completes still include the account ID when sent.
        account_id = self.get_account_id(get_boto3_session())
        if account_id:
            self.update_common_details({"accountId": account_id})

        while True:
            # Blocks until we get a new entry in the queue
            event_data: Optional[TelemetryEvent] = self.event_queue.get()
            # We've received the shutdown signal
            if event_data is None:
                return

            headers = {"Accept": "application-json", "Content-Type": "application-json"}
            try:
                # Merge _common_details into the per-event details at send
                # time (not enqueue time) so late-resolved fields like
                # accountId are included.
                details = {**event_data.event_details, **self._common_details}
                request_body = {
                    "BatchId": str(uuid.uuid4()),
                    "RumEvents": [
                        {
                            "details": str(json.dumps(details)),
                            "id": str(uuid.uuid4()),
                            "metadata": str(json.dumps(self._system_metadata)),
                            "timestamp": int(datetime.now().timestamp()),
                            "type": event_data.event_type,
                        },
                    ],
                    "UserDetails": {"sessionId": self.session_id, "userId": self.telemetry_id},
                }
                request_body_encoded = str(json.dumps(request_body)).encode("utf-8")
            except Exception as exc:
                logger.debug(f"Failed to serialize telemetry data. {str(exc)}")
                continue

            req = request.Request(url=self.endpoint, data=request_body_encoded, headers=headers)
            try:
                logger.debug("Sending telemetry data: %s", request_body)
                self._send_request(req)
            except Exception as exc:
                # Swallow any kind of uncaught exception and stop sending telemetry
                logger.debug(f"Error received from service. {str(exc)}")
                return
            self.event_queue.task_done()

    def _put_telemetry_record(self, event: TelemetryEvent) -> None:
        if not self._initialized or self.telemetry_opted_out:
            return
        try:
            self.event_queue.put_nowait(event)
        except Full:
            # Silently swallow the error if the event queue is full (due to throttling of the service)
            pass

    def record_vfs_mounting(self, successfully_mounted: bool):
        details: Dict[str, Any] = {"successfully_mounted": successfully_mounted}
        event_type = "com.amazon.rum.deadline.job_attachments.vfs_mount"
        self.record_event(event_type=event_type, event_details=details, from_gui=False)

    def _record_summary_statistics(
        self, event_type: str, summary: SummaryStatistics, from_gui: bool
    ):
        details: Dict[str, Any] = asdict(summary)
        self.record_event(event_type=event_type, event_details=details, from_gui=from_gui)

    def record_hashing_summary(self, summary: SummaryStatistics, *, from_gui: bool = False):
        self._record_summary_statistics(
            "com.amazon.rum.deadline.job_attachments.hashing_summary", summary, from_gui
        )

    def record_upload_summary(self, summary: SummaryStatistics, *, from_gui: bool = False):
        self._record_summary_statistics(
            "com.amazon.rum.deadline.job_attachments.upload_summary", summary, from_gui
        )

    def record_error(
        self, event_details: Dict[str, Any], exception_type: str, from_gui: bool = False
    ):
        event_details["exception_type"] = exception_type
        # Possibility to add stack trace here
        self.record_event("com.amazon.rum.deadline.error", event_details, from_gui=from_gui)

    @_swallow_exceptions
    def record_event(
        self, event_type: str, event_details: Dict[str, Any], *, from_gui: bool = False
    ):
        event_details["usage_mode"] = "GUI" if from_gui else "CLI"
        self._put_telemetry_record(
            TelemetryEvent(
                event_type=event_type,
                event_details=event_details,
            )
        )

    @lru_cache
    def get_account_id(self, boto3_session) -> Optional[str]:
        """Best-effort AWS account ID lookup for telemetry, cached per
        (client, session) so it runs at most once per telemetry client
        instance.

        Prefers ``session.get_credentials().account_id`` (populated for free
        by SSO, AssumeRole, IMDS/ECS, or a ``credential_process`` that emits
        ``AccountId``). Deadline Cloud monitor delivers its credentials
        through ``credential_process`` (see ``_get_boto3_session_for_profile``
        in ``_session.py``), so if DCM's process output includes
        ``AccountId`` this fast path covers the common monitor user flow
        without an STS call. Falls back to ``sts:GetCallerIdentity`` with a
        short timeout, returning ``None`` on any failure so users on
        restricted networks without STS access can still run the CLI.
        """
        try:
            credentials = boto3_session.get_credentials()
            account_id = getattr(credentials, "account_id", None) if credentials else None
            if account_id:
                return account_id
            # Short-timeout best-effort fallback; runs on the telemetry background
            # thread so blocking is fine.
            sts = boto3_session.client(
                "sts",
                config=BotocoreConfig(
                    connect_timeout=2, read_timeout=2, retries={"max_attempts": 1}
                ),
            )
            return sts.get_caller_identity()["Account"]
        except Exception:
            logger.debug("Could not resolve account ID for telemetry", exc_info=True)
            return None

    def update_common_details(self, details: Dict[str, Any]):
        """Updates the dict of common data that is included in every telemetry request."""
        self._common_details.update(details)


def get_telemetry_client(
    package_name: str, package_ver: str, config: Optional[ConfigParser] = None
) -> TelemetryClient:
    """
    Retrieves the cached telemetry client, lazy-loading the first time this is called.
    :param package_name: Base package name to associate data by.
    :param package_ver: Base package version to associate data by.
    :param config: Optional configuration to use for the client. Loads defaults if not given.
    :return: Telemetry client to make requests with.
    """
    global __cached_telemetry_clients
    cached = __cached_telemetry_clients.get(package_name)
    if not cached:
        cached = TelemetryClient(
            package_name=package_name,
            package_ver=package_ver,
            config=config,
        )
        __cached_telemetry_clients[package_name] = cached
    elif not cached.is_initialized:
        cached.initialize(config=config)

    return cached


def get_deadline_cloud_library_telemetry_client(
    config: Optional[ConfigParser] = None,
) -> TelemetryClient:
    """
    Retrieves the cached telemetry client, specifying the Deadline Cloud Client Library's package information.
    :param config: Optional configuration to use for the client. Loads defaults if not given.
    :return: Telemetry client to make requests with.
    """
    return get_telemetry_client("deadline-cloud-library", version, config=config)


def record_success_fail_telemetry_event(**decorator_kwargs: Any) -> Callable[[F], F]:
    """
    Decorator to try catch a function. Sends a success / fail telemetry event.
    :param ** Python variable arguments. See https://docs.python.org/3/glossary.html#term-parameter.
    """

    def inner(function: F) -> F:
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            """
            Wrapper to try-catch a function for telemetry
            :param * Python variable argument. See https://docs.python.org/3/glossary.html#term-parameter
            :param ** Python variable argument. See https://docs.python.org/3/glossary.html#term-parameter
            """
            success: bool = False
            try:
                result = function(*args, **kwargs)
                success = True
                return result
            finally:
                event_name = decorator_kwargs.get("metric_name", function.__name__)

                event_details: dict = decorator_kwargs.get("event_details", {})
                event_details["is_success"] = success
                raised_exception = sys.exc_info()[1]
                if raised_exception is not None:
                    event_details["exception_type"] = type(raised_exception).__name__

                get_deadline_cloud_library_telemetry_client().record_event(
                    event_type=f"com.amazon.rum.deadline.{event_name}",
                    event_details=event_details,
                )

        wrapper.__doc__ = function.__doc__
        return cast(F, wrapper)

    return inner


def record_function_latency_telemetry_event(**decorator_kwargs: Any) -> Callable[[F], F]:
    """
    Decorator to time a function. Sends a latency telemetry event.
    :param ** Python variable arguments. See https://docs.python.org/3/glossary.html#term-parameter.
    """

    def inner(function: F) -> F:
        @wraps(function)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            start_t = time.perf_counter_ns()
            ret_val = function(*args, **kwargs)
            end_t = time.perf_counter_ns()

            latency = end_t - start_t

            event_name = decorator_kwargs.get("metric_name", function.__name__)
            get_deadline_cloud_library_telemetry_client().record_event(
                event_type="com.amazon.rum.deadline.latency",
                event_details={"latency": latency, "function_call": event_name},
            )

            return ret_val

        return cast(F, wrapper)

    return inner

File: crates/deadline-common/src/telemetry.rs
//! Telemetry client — background event sender for Deadline Cloud.
//!
//! Matches Python's `_telemetry.py`: std::thread + std::sync::mpsc + ureq.
//! No async, no tokio — just a background OS thread doing blocking HTTP.

use deadline_config::config_file;
use deadline_config::ini::IniConfig;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::sync::mpsc::{self, SyncSender};
use std::thread;
use uuid::Uuid;

const MAX_QUEUE_SIZE: usize = 25;
const BASE_TIME: f64 = 0.5;
const MAX_BACKOFF_SECONDS: f64 = 10.0;
const MAX_RETRY_ATTEMPTS: u32 = 4;
const ENDPOINT_PREFIX: &str = "management.";

pub struct TelemetryEvent {
    pub event_type: String,
    pub event_details: HashMap<String, Value>,
}

pub struct TelemetryClient {
    sender: Option<SyncSender<TelemetryEvent>>,
    thread_handle: Option<thread::JoinHandle<()>>,
    initialized: bool,
    opted_out: bool,
    session_id: String,
    telemetry_id: String,
    common_details: HashMap<String, Value>,
    system_metadata: HashMap<String, Value>,
}

impl TelemetryClient {
    pub fn new(
        package_name: &str,
        package_ver: &str,
        config: Option<&IniConfig>,
    ) -> Self {
        let ver = truncate_version(package_ver);
        let telemetry_id = get_or_create_identifier(config);
        let mut common_details = HashMap::new();
        if package_name != "deadline-cloud-library" {
            common_details.insert(
                "deadline-cloud-version".into(),
                Value::String(env!("CARGO_PKG_VERSION").to_string()),
            );
        }

        let mut client = Self {
            sender: None,
            thread_handle: None,
            initialized: false,
            opted_out: false,
            session_id: Uuid::new_v4().to_string(),
            telemetry_id,
            common_details,
            system_metadata: build_system_metadata(package_name, &ver),
        };
        client.opted_out = resolve_opt_out(config);
        client
    }

    /// Start the background sender thread. Call after AWS config is available.
    pub fn initialize(&mut self, endpoint_url: &str, config: Option<&IniConfig>) {
        if self.opted_out {
            return;
        }

        let endpoint = prefix_endpoint(
            &format!("{endpoint_url}/2023-10-12/telemetry"),
            ENDPOINT_PREFIX,
        );

        // Add user_id and monitor_id if available
        if let Some(c) = config {
            if let Ok(uid) = config_file::get_setting_with_config("defaults.user_id", c) {
                if !uid.is_empty() {
                    self.system_metadata.insert("user_id".into(), Value::String(uid));
                }
            }
        }

        let (tx, rx) = mpsc::sync_channel::<TelemetryEvent>(MAX_QUEUE_SIZE);
        let session_id = self.session_id.clone();
        let telemetry_id = self.telemetry_id.clone();
        let system_metadata = self.system_metadata.clone();

        let handle = thread::spawn(move || {
            process_event_queue(rx, &endpoint, &session_id, &telemetry_id, &system_metadata);
        });

        self.sender = Some(tx);
        self.thread_handle = Some(handle);
        self.initialized = true;
    }

    pub fn is_initialized(&self) -> bool {
        self.initialized
    }

    pub fn record_event(
        &self,
        event_type: &str,
        mut event_details: HashMap<String, Value>,
        from_gui: bool,
    ) {
        if !self.initialized || self.opted_out {
            return;
        }
        for (k, v) in &self.common_details {
            event_details.entry(k.clone()).or_insert_with(|| v.clone());
        }
        event_details.insert(
            "usage_mode".into(),
            Value::String(if from_gui { "GUI" } else { "CLI" }.into()),
        );
        if let Some(ref tx) = self.sender {
            let _ = tx.try_send(TelemetryEvent {
                event_type: event_type.to_string(),
                event_details,
            });
        }
    }

    pub fn update_common_details(&mut self, details: HashMap<String, Value>) {
        self.common_details.extend(details);
    }
}

impl Drop for TelemetryClient {
    fn drop(&mut self) {
        // Drop the sender to signal the background thread to exit,
        // then join to ensure all queued events are sent.
        self.sender.take();
        if let Some(handle) = self.thread_handle.take() {
            let _ = handle.join();
        }
    }
}

// ---------------------------------------------------------------------------
// Background thread
// ---------------------------------------------------------------------------

fn process_event_queue(
    rx: mpsc::Receiver<TelemetryEvent>,
    endpoint: &str,
    session_id: &str,
    telemetry_id: &str,
    system_metadata: &HashMap<String, Value>,
) {
    while let Ok(event) = rx.recv() {
        let body = json!({
            "BatchId": Uuid::new_v4().to_string(),
            "RumEvents": [{
                "details": serde_json::to_string(&event.event_details).unwrap_or_default(),
                "id": Uuid::new_v4().to_string(),
                "metadata": serde_json::to_string(system_metadata).unwrap_or_default(),
                "timestamp": std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap_or_default()
                    .as_secs(),
                "type": event.event_type,
            }],
            "UserDetails": {
                "sessionId": session_id,
                "userId": telemetry_id,
            },
        });

        if let Err(e) = send_with_retry(endpoint, &body) {
            log::debug!("Telemetry send failed: {e}");
            return; // Stop processing on unrecoverable error (matches Python)
        }
    }
}

fn send_with_retry(endpoint: &str, body: &Value) -> Result<(), String> {
    let body_str = body.to_string();
    for attempt in 0..MAX_RETRY_ATTEMPTS {
        match ureq::post(endpoint)
            .header("Accept", "application-json")
            .header("Content-Type", "application-json")
            .send(body_str.as_bytes())
        {
            Ok(_) => return Ok(()),
            Err(ureq::Error::StatusCode(429)) | Err(ureq::Error::StatusCode(500)) => {
                if attempt + 1 >= MAX_RETRY_ATTEMPTS {
                    return Err("Max retries reached sending telemetry".into());
                }
                let backoff = rand_backoff(attempt + 1);
                thread::sleep(std::time::Duration::from_secs_f64(backoff));
            }
            Err(e) => return Err(e.to_string()),
        }
    }
    Err("Max retries reached".into())
}

fn rand_backoff(attempt: u32) -> f64 {
    let max = MAX_BACKOFF_SECONDS.min(BASE_TIME * 2.0_f64.powi(attempt as i32));
    // Simple pseudo-random: use thread ID + time as seed
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .subsec_nanos() as f64;
    (nanos % 1000.0) / 1000.0 * max
}

// ---------------------------------------------------------------------------
// Helpers (public for testing)
// ---------------------------------------------------------------------------

/// Truncate version to first 3 components: "1.2.3.4.5" → "1.2.3"
pub fn truncate_version(ver: &str) -> String {
    ver.splitn(4, '.').take(3).collect::<Vec<_>>().join(".")
}

/// Insert prefix after "https://": "https://example.com" → "https://management.example.com"
pub fn prefix_endpoint(endpoint: &str, prefix: &str) -> String {
    if let Some(rest) = endpoint.strip_prefix("https://") {
        format!("https://{prefix}{rest}")
    } else {
        endpoint.to_string()
    }
}

/// Check opt-out: env var supersedes config.
pub fn resolve_opt_out(config: Option<&IniConfig>) -> bool {
    let env_val = std::env::var("DEADLINE_CLOUD_TELEMETRY_OPT_OUT").unwrap_or_default();
    if !env_val.is_empty() {
        return config_file::str2bool(&env_val).unwrap_or(false);
    }
    let val = match config {
        Some(c) => config_file::get_setting_with_config("telemetry.opt_out", c)
            .unwrap_or_default(),
        None => config_file::get_setting("telemetry.opt_out").unwrap_or_default(),
    };
    config_file::str2bool(&val).unwrap_or(false)
}

/// Validate existing identifier or generate a new UUID4.
pub fn validate_or_generate_identifier(existing: Option<&str>) -> String {
    if let Some(id) = existing {
        if Uuid::parse_str(id).is_ok() {
            return id.to_string();
        }
    }
    Uuid::new_v4().to_string()
}

fn get_or_create_identifier(config: Option<&IniConfig>) -> String {
    let existing = match config {
        Some(c) => config_file::get_setting_with_config("telemetry.identifier", c).ok(),
        None => config_file::get_setting("telemetry.identifier").ok(),
    };
    let id = validate_or_generate_identifier(existing.as_deref());
    if existing.as_deref() != Some(&id) {
        // Save the new identifier — best effort
        let _ = config_file::set_setting("telemetry.identifier", &id);
    }
    id
}

fn build_system_metadata(package_name: &str, package_ver: &str) -> HashMap<String, Value> {
    let mut m = HashMap::new();
    m.insert("service".into(), Value::String(package_name.into()));
    m.insert("version".into(), Value::String(package_ver.into()));
    m.insert("osName".into(), Value::String(
        if cfg!(target_os = "macos") { "macOS" }
        else if cfg!(target_os = "windows") { "Windows" }
        else { "Linux" }
        .into(),
    ));
    m
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_case::test_case;

    // --- Version truncation (pure function) ---

    #[test_case("1.2.3.4.5", "1.2.3" ; "five components truncated to three")]
    #[test_case("1.2.3", "1.2.3" ; "three components unchanged")]
    #[test_case("1.2", "1.2" ; "two components unchanged")]
    #[test_case("1", "1" ; "one component unchanged")]
    fn truncate_version_to_three_components(input: &str, expected: &str) {
        assert_eq!(truncate_version(input), expected);
    }

    // --- Endpoint prefixing (pure function) ---

    #[test_case(
        "https://deadline.us-west-2.amazonaws.com",
        "management.",
        "https://management.deadline.us-west-2.amazonaws.com"
        ; "inserts prefix after https"
    )]
    #[test_case(
        "http://localhost:8080",
        "management.",
        "http://localhost:8080"
        ; "non-https unchanged"
    )]
    fn prefix_endpoint_inserts_after_https(endpoint: &str, prefix: &str, expected: &str) {
        assert_eq!(prefix_endpoint(endpoint, prefix), expected);
    }

    // --- Telemetry identifier (pure function) ---

    #[test]
    fn valid_uuid4_identifier_is_preserved() {
        let id = "550e8400-e29b-41d4-a716-446655440000";
        assert_eq!(validate_or_generate_identifier(Some(id)), id);
    }

    #[test]
    fn invalid_identifier_generates_new_uuid4() {
        let result = validate_or_generate_identifier(Some("not-a-uuid"));
        assert_ne!(result, "not-a-uuid");
        Uuid::parse_str(&result).expect("should be valid UUID");
    }

    #[test]
    fn empty_identifier_generates_new_uuid4() {
        let result = validate_or_generate_identifier(None);
        Uuid::parse_str(&result).expect("should be valid UUID");
    }
}
